### PR TITLE
Add GRLC endpoint for replaced_by annotations

### LINKED ISSUE #77: Add GRLC query for an API that retrieves "relacedBy" annotations
We need a GRLC query for an API that retrieves "relacedBy" annotations for terms to provide to our stakeholders. For now the API should, ideally, be resilient to:

- using alt_id instead of replacedBy
- using CURIE strings instead of IRI

But ideally, we should push for all our ontologies to fix these inconsistent representations at source.

### PR DESCRIPTION
**Pull request by anitacaron**
Fixes #77

### CODE REVIEW

**Review by matentzn — APPROVED** [review]
This does not deal with the fact that some ontologies still use `CU:RIE^^xsd:string` syntax for replacements, but we should fix this on ontology level. Thanks!

**Comment by matentzn** [comment]
maybe get another review from @ubyndr

**Comment by anitacaron** [comment]
I tested [here](https://api.triplydb.com/s/_JnZEFGrD), and it also lists this case.

**Comment by matentzn** [comment]
I dont doubt the query returns all the information, but what I am saying is that an API user will be confused about the fact that some replacements are strings and others IRIs..

In Fact, its probably better to add an isIRI(?replacement_id) filter to the queries to avoid this, and make sure that we fix the ontologies..

**Comment by anitacaron** [comment]
I noticed that using `alt_id` gives the IRI and `replaced_by` gives in string.

**Comment by matentzn** [comment]
> I noticed that using alt_id gives the IRI and replaced_by gives in string.

We should prioritise this _not_ being the case for all our OWL releases.

File: sparql/obsolete_replaced-by_term.rq [hunk 1]
```diff
@@ -0,0 +1,22 @@
+#+ summary: Find replaced by term in obsolete term
+#+ description: Enter an obsolete term and return the term replaced by it 
+
+PREFIX xsd: <http://www.w3.org/2001/XMLSchema#>
+PREFIX owl: <http://www.w3.org/2002/07/owl#>
+PREFIX rdf: <http://www.w3.org/1999/02/22-rdf-syntax-ns#>
+PREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>
```
Comment (ubyndr) [inline sparql/obsolete_replaced-by_term.rq:7]: seems like there is no use for this prefix.

File: sparql/obsolete_replaced-by_terms_by_ontology.rq [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+      replaced_by: ?replaced_by; 
+      rdfs:isDefinedBy ?_ontology_iri .
+  }
+  UNION
+  {
+    ?replaced_by rdf:type owl:Class;
+      oboinowl:hasAlternativeId ?obsolete_class .
+    ?obsolete_class rdfs:isDefinedBy ?_ontology_iri .
+  }
+  FILTER(isIRI(?replaced_by))
```
Comment (ubyndr) [inline sparql/obsolete_replaced-by_terms_by_ontology.rq:52]: just a question, is it possible to have a string or anything other than an IRI as range of replaced_by: ?

**Review by ubyndr — APPROVED** [review]
looks good

File: sparql/obsolete_replaced-by_terms_by_ontology.rq (hunk 2 above)
Comment (ubyndr) [reply sparql/obsolete_replaced-by_terms_by_ontology.rq:52]: I withdraw my question 😃

Comment (anitacaron) [reply sparql/obsolete_replaced-by_terms_by_ontology.rq:52]: 😄

File: sparql/obsolete_replaced-by_term.rq [hunk 3]
```diff
@@ -0,0 +1,21 @@
+#+ summary: Find replaced by term in obsolete term
+#+ description: Enter an obsolete term and return the term replaced by it
```
Comment (balhoff) [inline sparql/obsolete_replaced-by_term.rq:2]: Change to:
`#+ description: Enter an obsolete term and return the term that replaces it`

File: sparql/obsolete_replaced-by_terms_by_ontology.rq [hunk 4]
```diff
@@ -0,0 +1,53 @@
+#+ summary: Find replaced terms by ontology
+#+ description: Choose an ontology and get back list of all obsolete terms with their terms replaced by it
```
Comment (balhoff) [inline sparql/obsolete_replaced-by_terms_by_ontology.rq:2]: Change to:
`#+ description: Choose an ontology and get back a list of all obsolete terms with the terms that replace them`

**Review by balhoff — CHANGES_REQUESTED** [review]

File: sparql/obsolete_replaced-by_term.rq [hunk 5]
```diff
@@ -0,0 +1,21 @@
...
+PREFIX xsd: <http://www.w3.org/2001/XMLSchema#>
+PREFIX owl: <http://www.w3.org/2002/07/owl#>
+PREFIX rdf: <http://www.w3.org/1999/02/22-rdf-syntax-ns#>
+PREFIX replaced_by: <http://purl.obolibrary.org/obo/IAO_0100001>
+SELECT ?obsolete_class ?replaced_by
+WHERE {
+  {
+    ?_obsolete_class_iri rdf:type owl:Class; 
+      owl:deprecated "true"^^xsd:boolean; 
+      replaced_by: ?replaced_by .
```
Comment (balhoff) [inline sparql/obsolete_replaced-by_term.rq:13]: You could BIND another variable as something like `?replaced_by_iri` and use an `IF` to convert CURIE strings to IRIs assuming they are OBO-style CURIEs.

File: sparql/obsolete_replaced-by_terms_by_ontology.rq [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+      replaced_by: ?replaced_by; 
+      rdfs:isDefinedBy ?_ontology_iri .
+  }
+  UNION
+  {
+    ?replaced_by rdf:type owl:Class;
+      oboinowl:hasAlternativeId ?obsolete_class .
+    ?obsolete_class rdfs:isDefinedBy ?_ontology_iri .
+  }
+  BIND (IF(isLiteral(?replaced_by), IRI(CONCAT("http://purl.obolibrary.org/obo/", STRBEFORE(?replaced_by_iri, ":"), "_", STRAFTER(?replaced_by_iri, ":"))), ?replaced_by) AS ?replaced_by_iri)
```
Comment (matentzn) [inline sparql/obsolete_replaced-by_terms_by_ontology.rq:54]:
```suggestion
  BIND (IF(isLiteral(?replaced_by), IRI(CONCAT("http://purl.obolibrary.org/obo/", STRBEFORE(?replaced_by, ":"), "_", STRAFTER(?replaced_by, ":"))), ?replaced_by) AS ?replaced_by_iri)

```

**Review by matentzn — CHANGES_REQUESTED** [review]
I dont know GARLC well enough to judge this! The only concern I have (aside from the typo I found) is that the SPARQL query assumes that the replaced by value is either a valid IRI or a valid OBO CURIE, while it is possible that the replaced by is:

1. A non-OBO CURIE (you cant do anything about that)
2. Neither an IRI nor a CURIE (too cumbersome to do anything about that).

File: sparql/obsolete_replaced-by_terms_by_ontology.rq (hunk 6 above)
Comment (anitacaron) [reply sparql/obsolete_replaced-by_terms_by_ontology.rq:54]: GRLC works the same as the SPARQL query. You'll get the wrong URL for the first case, and it won't show any results for the second case.

**Review by matentzn — APPROVED** [review]

**Pull request merged**